planner/core/plan_cache_param.go

```diff
@@ -0,0 +1,163 @@
+}
+
+func (pr *paramReplacer) Reset() { pr.params = nil }
+
+// ParameterizeAST parameterizes this StmtNode.
+// e.g. `select * from t where a<10 and b<23` --> `select * from t where a<? and b<?`, [10, 23]
+// The returned vals are used to restore this AST if encounter some errors, see RestoreASTWithParams.
+func ParameterizeAST(sctx sessionctx.Context, stmt ast.StmtNode) (paramSQL string, params []*driver.ValueExpr, ok bool, err error) {
+	if !Available4GeneralPlanCache(sctx, stmt) { // not support
+		return "", nil, false, nil
+	}
+
+	pr := paramReplacerPool.Get().(*paramReplacer)
+	pCtx := paramCtxPool.Get().(*format.RestoreCtx)
+	defer func() {
+		pr.Reset()
+		paramReplacerPool.Put(pr)
+		pCtx.In.(*strings.Builder).Reset()
+		paramCtxPool.Put(pCtx)
+	}()
+	stmt.Accept(pr)
+	if err := stmt.Restore(pCtx); err != nil {
+		err = RestoreASTWithParams(sctx, stmt, pr.params)
+		return "", nil, false, err
+	}
+
+	paramSQL = pCtx.In.(*strings.Builder).String()
+	params = pr.params
+	ok = true
+	return
+}
+
+type paramRestorer struct {
+	params []*driver.ValueExpr
+	err    error
+}
+
+func (pr *paramRestorer) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
+	switch n := in.(type) {
+	case *driver.ParamMarkerExpr:
+		if n.Offset >= len(pr.params) {
+			pr.err = errors.New("TODO")
+			return nil, true
+		}
+		// offset is used as order in general plan cache.
+		return pr.params[n.Offset], true
+	}
+	if pr.err != nil {
+		return nil, true
+	}
+	return in, false
+}
+
+func (pr *paramRestorer) Leave(in ast.Node) (out ast.Node, ok bool) {
+	return in, true
+}
+
+func (pr *paramRestorer) Reset() {
+	pr.params = nil
+	pr.err = nil
+}
+
+// RestoreASTWithParams restore this parameterized AST with specific parameters.
+// e.g. `select * from t where a<? and b<?`, [10, 23] --> `select * from t where a<10 and b<23`.
+func RestoreASTWithParams(_ sessionctx.Context, stmt ast.StmtNode, params []*driver.ValueExpr) error {
+	pr := paramRestorerPool.Get().(*paramRestorer)
+	defer func() {
+		pr.Reset()
+		paramRestorerPool.Put(pr)
+	}()
+	pr.params = params
+	stmt.Accept(pr)
+	return pr.err
+}
+
+// Available4GeneralPlanCache returns whether this AST is available for general plan cache.
+func Available4GeneralPlanCache(sctx sessionctx.Context, stmt ast.StmtNode) bool {
+	// Enter only support: select {col} from {single-table} where {cond} and {cond} ...
+	// 	and {cond} = {col} {op} {val}
+	// 	and {op} = >, <, =
+	n, ok := stmt.(*ast.SelectStmt)
+	if !ok { // TODO: support DML statements
+		return false
+	}
+	if len(n.TableHints) > 0 || // hints
+		n.Distinct || n.GroupBy != nil || n.Having != nil || // agg
+		n.WindowSpecs != nil || // window function
+		n.OrderBy != nil || n.Limit != nil || // order / limit
+		n.LockInfo != nil || n.SelectIntoOpt != nil { // lock info
+		return false
+	}
+	if n.From == nil || n.From.TableRefs == nil { // not from a general table
+		return false
+	}
+	if n.From.TableRefs.Right != nil { // no join
+		return false
+	}
```

Comment:
Maybe should add this part 
```
switch x := n.From.TableRefs.Left.(type) {
 		case *ast.TableSource:
 			_, isTableName := x.Source.(*ast.TableName)
 			if !isTableName {
 				return false
 			}
 		}
```
For example, subquery or something else can also match this part. Please add some test cases for this.

Reply:
How about letting us improve it and do more detailed checks on the following several PRs? The main purpose of this PR is to introduce `ParameterizeAST` and `RestoreASTWithParams`.

Reply:
Emm, I removed this function. Let's implement this checker on the next PR.